File: pkg/sunbeam/manifest.go
package sunbeam

type Manifest struct {
	Title       string    `json:"title"`
	Description string    `json:"description,omitempty"`
	Root        []Action  `json:"root,omitempty"`
	Commands    []Command `json:"commands"`
}

type Command struct {
	Name        string      `json:"name"`
	Description string      `json:"description,omitempty"`
	Params      []Param     `json:"params,omitempty"`
	Mode        CommandMode `json:"mode,omitempty"`
}

type Platfom string

const (
	PlatformLinux Platfom = "linux"
	PlatformMac   Platfom = "macos"
)

type Requirement struct {
	Name string `json:"name"`
	Link string `json:"link,omitempty"`
}

type CommandMode string

const (
	CommandModeSearch CommandMode = "search"
	CommandModeFilter CommandMode = "filter"
	CommandModeDetail CommandMode = "detail"
	CommandModeSilent CommandMode = "silent"
)

type InputType string

const (
	InputString  InputType = "string"
	InputBoolean InputType = "boolean"
	InputNumber  InputType = "number"
)

type Param struct {
	Type        InputType `json:"type"`
	Name        string    `json:"name"`
	Description string    `json:"description,omitempty"`
	Optional    bool      `json:"optional,omitempty"`
	Default     any       `json:"default,omitempty"`
}

File: pkg/python/sunbeam/manifest.py
from typing import TypedDict, NotRequired, Literal, Sequence


class TextInput(TypedDict):
    default: NotRequired[str]
    placeholder: NotRequired[str]


class TextAreaInput(TypedDict):
    default: NotRequired[str]
    placeholder: NotRequired[str]


class CheckboxInput(TypedDict):
    default: NotRequired[bool]


class PasswordInput(TypedDict):
    placeholder: NotRequired[str]


class NumberInput(TypedDict):
    default: NotRequired[int]
    placeholder: NotRequired[str]


class Input(TypedDict):
    label: str
    type: str
    name: str
    optional: NotRequired[bool]
    text: NotRequired[TextInput]
    textarea: NotRequired[TextAreaInput]
    checkbox: NotRequired[CheckboxInput]
    password: NotRequired[PasswordInput]
    number: NotRequired[NumberInput]


CommandMode = Literal["silent", "tty", "filter", "search", "detail"]


class Command(TypedDict):
    name: str
    title: str
    mode: CommandMode
    params: NotRequired[list[Input]]


class Manifest(TypedDict):
    title: str
    description: NotRequired[str]
    preferences: NotRequired[list[Input]]
    commands: NotRequired[list[Command]]
